### PR TITLE
set Vite's `publicDir` and `base` options

### PR DESCRIPTION
**Pull request by benmccann**
Benefits:
1. A lot less code in SvelteKit
1. `publicDir` and `base` are now defined for Vite plugins (should help @userquin with his work supporting SvelteKit in `vite-plugin-pwa`)
1. The Vite CLI now correctly includes the base URL, so you can Ctrl+click it in the terminal to be taken to the correct location

I would suggest closing https://github.com/sveltejs/kit/issues/5115 with a suggestion that the issue be filed in the Vite repo

### CODE REVIEW

**Comment by userquin** [comment]
@benmccann it is the webmanifest file, not the sw: pwa plugin will emit `manifest.webmanifest` asset in generateBundle but kit will copy it to _app folder.

**Comment by Rich-Harris** [comment]
With this approach, the static assets get copied into `.svelte-kit/output/client`. Presumably we'd need to remove `writeStatic` from the adapter API if we did this

**Comment by userquin** [comment]
> With this approach, the static assets get copied into `.svelte-kit/output/client`. Presumably we'd need to remove `writeStatic` from the adapter API if we did this

Why not just replace static with public folder? I don't need the assets to be written in the output folder.

I only need the publicDir to list the assets matching a globPattern

File: packages/kit/test/apps/options-2/test/test.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ test.describe.configure({ mode: 'parallel' });
 
 test.describe('paths.base', () => {
 	test('serves /basepath', async ({ page }) => {
-		await page.goto('/basepath');
+		await page.goto('/basepath/');
```
Comment (Rich-Harris) [inline packages/kit/test/apps/options-2/test/test.js:10]: what's the reason for this change?

**Comment by Conduitry** [comment]
This will also need changesets for all the adapters, correct?

**Comment by Rich-Harris** [comment]
It would, yeah

File: packages/kit/test/apps/options-2/test/test.js (hunk 1 above)
Comment (benmccann) [reply packages/kit/test/apps/options-2/test/test.js:10]: Vite only serves the index page with a trailing slash

Comment (benmccann) [reply packages/kit/test/apps/options-2/test/test.js:10]: Related issue: https://github.com/vitejs/vite/issues/8770

File: packages/kit/src/core/prerender/prerender.js [hunk 2]
```diff
@@ -109,7 +109,7 @@ export async function prerender({ config, entries, files, log }) {
 		if (seen.has(decoded)) return;
 		seen.add(decoded);
 
-		const file = decoded.slice(config.paths.base.length + 1);
+		const file = decoded;
```
Comment (Rich-Harris) [inline packages/kit/src/core/prerender/prerender.js:112]: why did you remove this? no wonder the adapter-static tests are failing 😜

```suggestion
		const file = decoded.slice(config.paths.base.length + 1);
```

File: packages/kit/src/core/prerender/prerender.js [hunk 3]
```diff
@@ -259,10 +254,10 @@ export async function prerender({ config, entries, files, log }) {
 		for (const entry of config.prerender.entries) {
 			if (entry === '*') {
 				for (const entry of entries) {
-					enqueue(null, config.paths.base + entry); // TODO can we pre-normalize these?
```
Comment (Rich-Harris) [inline packages/kit/src/core/prerender/prerender.js:262]: I don't understand what's going on with the removal of `config.paths.base`. Surely it needs to be there?

File: packages/kit/test/apps/options-2/test/test.js (hunk 1 above)
Comment (Rich-Harris) [reply packages/kit/test/apps/options-2/test/test.js:10]: okay, but the path is incorrect

File: packages/kit/src/core/prerender/prerender.js (hunk 3 above)
Comment (Rich-Harris) [reply packages/kit/src/core/prerender/prerender.js:262]: also, not sure why you got rid of the comment. it's still a TODO

File: packages/kit/src/core/prerender/prerender.js (hunk 2 above)
Comment (benmccann) [reply packages/kit/src/core/prerender/prerender.js:112]: Vite's `baseMiddleware` strips out the `base` path from the URL
https://github.com/vitejs/vite/blob/dfaeb2b7dc39822bf00fc97802b758b17129c37b/packages/vite/src/node/server/index.ts#L513
https://github.com/vitejs/vite/blob/dfaeb2b7dc39822bf00fc97802b758b17129c37b/packages/vite/src/node/server/middlewares/base.ts#L17

File: packages/kit/src/core/prerender/prerender.js (hunk 3 above)
Comment (benmccann) [reply packages/kit/src/core/prerender/prerender.js:262]: I didn't understand the TODO. I thought it was referring to prepending the base

The base would not be needed as Vite's `baseMiddleware` strips out the `base` path from the URL as mentioned above https://github.com/sveltejs/kit/pull/5601#discussion_r924911591

File: packages/kit/test/apps/basics/test/server.test.js [hunk 4]
```diff
@@ -473,11 +473,6 @@ test.describe('Static files', () => {
 		expect(await r2.json()).toEqual({ works: true });
 	});
 
-	test('Filenames are case-sensitive', async ({ request }) => {
-		const response = await request.get('/static.JSON');
-		expect(response.status()).toBe(404);
-	});
```
Comment (Rich-Harris) [inline packages/kit/test/apps/basics/test/server.test.js:479]: why was this test deleted?

Comment (benmccann) [reply packages/kit/test/apps/basics/test/server.test.js]: it looks like it doesn't need to be actually. I just restored it

Comment (Rich-Harris) [reply packages/kit/test/apps/basics/test/server.test.js:479]: Right, but many servers do, so it's essential that we do too — if we can't get Vite to be strict about casing it's a showstopper for this PR

**Comment by benmccann** [comment]
Closing this for now as it doesn't seem feasible until maybe Vite 4 given current limitations in Vite
